### PR TITLE
Nix derivation and shell

### PR DESCRIPTION
**Pull request by BenSchZA**
Generated initial Nix derivation for building and creating a shell with all necessary dependencies for Linux and Mac systems (although Windows is also in the works).

This is still a work in progress, needs proper documentation. Adding PR to update on progress.

## Next steps
1. Documentation
2. Possibly generate actual cadCAD package as a Nix package for easy installation

### CODE REVIEW

**Comment by JEJodesty** [comment]
Congratulations! You've submitted our first PR :D !

**Comment by JEJodesty** [comment]
I'm publishing contribution rules this weekend and even then I consider it a work in progress
We have dedicated feature branches upstream. I will create a branch 'nix-shell' or 'nix_shell' this weekend so you could submit a PR to that branch. 

The PR ("->") path for features is as follows:
Option A: Ideal
fork/feature -> BlockScience/feature -> BlockScience/staging -> BlockScience/master
Option B: For Now
fork/feature -> BlockScience/staging -> BlockScience/master

I would recommend not naming branches with `/` because the left-hand side can be mistaken for a remote. All left-hand sides above are remotes.

Only Maintainers can submit PRs to `master`
Contributor PRs to `staging` can be submitted for now, but thats still being debated internally.
I'm thing of adding a buffer branch between down and upstream on upstream like `staging` but its still being discussed.

Anyway, thanks for your contribution
I'm just excited to have an initial PR!

**Comment by BenSchZA** [comment]
Thanks @JEJodesty ! Excited to submit the first PR, and hopefully it won't be the last :) Let me know once you've created that feature branch and I'll move the changes over there.

**Comment by JEJodesty** [comment]
@BenSchZA 
sorry for the long wait things go crazy over here, `nix-shell` is a branch so submit a PR to it 😃

**Comment by JEJodesty** [comment]
@BenSchZA 
added feature branch, sorry for the wait :( :)

**Comment by JEJodesty** [comment]
also im assuming this
https://nixos.org/nix/
?

**Comment by BenSchZA** [comment]
> also im assuming this
> https://nixos.org/nix/
> ?

That's correct @JEJodesty Once I'm happy the Nix derivation is working properly I might also add it to the Nixpkgs package repository. It really is a great way to manage packages for Linux/Mac :)

**Comment by mzargham** [comment]
@markusbkoch @JEJodesty I just came across this hanging pull request. Is there a reason it hasn't been merged?

**Comment by BenSchZA** [comment]
It's still a work in progress that I put on hold when the Python package was made public. I was initially using it to build the package locally, but I imagine it'll need some updates now.

-------- Original Message --------

**Comment by BenSchZA** [comment]
@JEJodesty I've updated the Nix package and the README. I think it's ready to be merged.

I've also started a PR to the Nixpkgs repository: https://github.com/BenSchZA/nixpkgs/blob/pythonPackage/cadCAD/pkgs/development/python-modules/cadCAD/default.nix

**Comment by BenSchZA** [comment]
I merged the cadCAD upstream master into my feature branch, might be worth doing the same with the BlockScience:nix-shell branch before we merge this in.

**Comment by JEJodesty** [comment]
> I merged the cadCAD upstream master into my feature branch, might be worth doing the same with the BlockScience:nix-shell branch before we merge this in.

**Comment by JEJodesty** [comment]
@BenSchZA 
Sorry this is taking long.
We're block for a complicated merge. May take some time.
But I have't lost track of this

**Comment by BenSchZA** [comment]
No problem @JEJodesty ! I heard there were some new features and performance improvements coming soon, looking forward to seeing them :)

**Comment by JEJodesty** [comment]
@BenSchZA 
check your telegram

File: default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{ pkgs ? import <nixpkgs> { }, stdenv ? pkgs.stdenv, fetchPypi ? pkgs.fetchPypi
+, pythonPkgs ? pkgs.python37Packages
+, buildPythonPackage ? pythonPkgs.buildPythonPackage }:
+
+buildPythonPackage rec {
+  pname = "cadCAD";
+  version = "0.3.1";
```
Comment (JEJodesty) [inline default.nix:7]: 0.4.15

File: shell.nix [hunk 2]
```diff
@@ -0,0 +1,4 @@
+with import <nixpkgs> { };
+
+(let cadCAD = pkgs.callPackage ./default.nix { inherit (pkgs) ; };
+in pkgs.python37.withPackages (ps: [ cadCAD ps.pandas ])).env
```
Comment (JEJodesty) [inline shell.nix:4]: python 3.6.8

File: default.nix [hunk 3]
```diff
@@ -0,0 +1,40 @@
+{ pkgs ? import <nixpkgs> { }, stdenv ? pkgs.stdenv, fetchPypi ? pkgs.fetchPypi
+, pythonPkgs ? pkgs.python37Packages
```
Comment (JEJodesty) [inline default.nix:2]: python 3.6.8

File: default.nix [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  #  inherit pname version;
+  #  sha256 = "4f2a4d39e4ea601b9ab42b2db08b5918a9538c168cff1c6895ae26646f3d73b1";
+  #};
+  # src = builtins.fetchGit {
+  #   url = "git@github.com:BlockScience/cadCAD.git";
+  #   ref = "master";
+  # };
+  src = ./.;
+
+  doCheck = false; # TODO: correct or disclude integration tests
```
Comment (JEJodesty) [inline default.nix:19]: remove ToDo

File: default.nix [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  #   ref = "master";
+  # };
+  src = ./.;
+
+  # In future, when tests are introduced, this can be updated to 'true'
+  doCheck = false;
+
+  buildInputs = [ ];
+  checkInputs = [ ];
+  propagatedBuildInputs = with pythonPkgs; [
```
Comment (JEJodesty) [inline default.nix:25]: check setup.py (core dependencies)
also requirements txt

**Comment by JEJodesty** [comment]
https://github.com/BlockScience/cadCAD/pull/59

**Pull request merged**